### PR TITLE
Migrate `components/post_view/post_flag_icon/post_flag_icon.tsx` from class to function component

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Migrate `components/post_view/post_flag_icon/post_flag_icon.tsx` from class to function component

#### Ticket Link
None

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx [hunk 1]
```diff
@@ -19,115 +19,98 @@ export type Actions = {
...
-    handlePress = (e: React.MouseEvent) => {
+const PostFlagIcon = ({
+    actions: {
+        flagPost,
+        unflagPost,
+    },
+    isFlagged,
+    postId,
+    location = Locations.CENTER,
+}: Props) => {
```
Comment (larkox) [inline webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:37]: Expanded props to better track unused props.

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx [hunk 2]
```diff
@@ -19,115 +19,98 @@ export type Actions = {
...
+    const buttonRef = useRef<HTMLButtonElement>(null);
+    const [a11yActive, setA11yActive] = useState(false);
+
+    const handlePress = useCallback((e: React.MouseEvent) => {
```
Comment (larkox) [inline webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:41]: `useCallback` to avoid passing a new function to children objects on every render.

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx [hunk 3]
```diff
@@ -19,115 +19,98 @@ export type Actions = {
...
-        this.setState({a11yActive: true});
+    const handleA11yActivateEvent = () => {
+        setA11yActive(true);
     };
 
-    handleA11yDeactivateEvent = () => {
-        this.setState({a11yActive: false});
+    const handleA11yDeactivateEvent = () => {
+        setA11yActive(false);
     };
```
Comment (larkox) [inline webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:57]: No need for `useCallback` since are only used on on mount/dismount

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx [hunk 4]
```diff
@@ -19,115 +19,98 @@ export type Actions = {
...
+                className='post-menu__item'
+                onClick={handlePress}
+            >
+                {flagIcon}
+            </button>
+        </OverlayTrigger>
+    );
+};
+
+export default React.memo(PostFlagIcon);
```
Comment (larkox) [inline webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:116]: `Rect.memo` since the component used to be a Pure Component.

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx [hunk 5]
```diff
@@ -19,115 +19,98 @@ export type Actions = {
...
         } else {
-            actions.flagPost(postId);
+            flagPost(postId);
         }
-    };
+    }, [flagPost, unflagPost, postId, isFlagged]);
 
-    handleA11yActivateEvent = () => {
-        this.setState({a11yActive: true});
+    const handleA11yActivateEvent = () => {
```
Comment (jespino) [inline webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:51]: I think this 2 functions should be defined inside the `useEffect`, that is the right scope for them, and there is not reason to have it outside.

**Review by jespino — CHANGES_REQUESTED** [review]
I have a tiny improvement suggestion, but otherwise looks good to me.

**Review by jespino — APPROVED** [review]
LGTM

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx [hunk 6]
```diff
@@ -19,115 +19,97 @@ export type Actions = {
...
-    handleA11yDeactivateEvent = () => {
-        this.setState({a11yActive: false});
-    };
+    useEffect(() => {
+        const handleA11yActivateEvent = () => {
+            setA11yActive(true);
+        };
+        const handleA11yDeactivateEvent = () => {
+            setA11yActive(false);
+        };
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:57]: Any reasons why these are anonymous functions stored in a variable instead of plain functions?

**Review by M-ZubairAhmed — APPROVED** [review]

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx [hunk 7]
```diff
@@ -19,115 +19,97 @@ export type Actions = {
     unflagPost: typeof unflagPost;
 }
 
-interface Props {
+type Props = {
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:22]: why the change here

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx (hunk 6 above)
Comment (larkox) [reply webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:57]: If you mean why use `const foo = () => {...}` instead of using just `() => {...}` where needed, it is because I want to send the same reference to both add and remove listener.

If you mean why use `const foo = () => {...}` instead of `func foo () {...}`, no reason at all. Not sure if there is any real difference between both options.

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx (hunk 7 above)
Comment (larkox) [reply webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:22]: I was receiving a weird typescript error (maybe just on local) where the index file was complaining with the following error: `Default export of the module has or is using private name 'Props'.`. Moving it to a type fixed it.

**Comment by larkox** [comment]
@M-ZubairAhmed The failing test seems to be non related to this. Any chance it is a flaky test?

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx (hunk 6 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:57]: yes i meant latter, there is no difference anyways. We don't want to set expectations for contributors to always use anonymous functions instead of plain functions.

File: webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx (hunk 7 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/post_view/post_flag_icon/post_flag_icon.tsx:22]: hmm thats interesting

**Comment by M-ZubairAhmed** [comment]
> @M-ZubairAhmed The failing test seems to be non related to this. Any chance it is a flaky test?

Yes this is a flaky test, happens with my PRs randomly sometimes as well

`src/components/confirm_modal_redux/confirm_modal_redux.test.tsx`

**Comment by larkox** [comment]
@yasserfaraazkhan QA Notes:

The `PostFlagIcon` is the icon to mark a post as saved. It appears in the PostOptions component, which is the options that appear when you hover over a post.

No functionality should have changed.

**Review by yasserfaraazkhan — APPROVED** [review]
